### PR TITLE
Adding tests to verify the crd generation fix

### PR DESCRIPTION
**Pull request by anupama2501**
## Issue: 
https://github.com/rancher/qa-tasks/issues/820
 
## Problem
In this PR, adding tests for the latest CRD generation fix. Currently the test cases for kubectl explain and kubectl validate are only applicable for roletemplates. But for the next release once we have all the changes in, we should update the code to ensure the other crd are validated as well.
 
## Solution
We do not jave kubectl apply or kubectl explain in our framework. Solving this by creating a job and using the job, we are executing into the pod and running the kubectl explain and kubectl apply commands. Once the job is either completed or errored out, we return the pod logs to the tests. 
 
## Testing
Jenkins log

### CODE REVIEW

File: tests/v2/validation/crdgeneration/crd_generation.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+			crdMap[groupProvisioning] = append(crdMap[groupProvisioning], crd)
+		} else if strings.Contains(crd, groupFleet) {
+			crdMap[groupFleet] = append(crdMap[groupFleet], crd)
+		} else {
+			crdMap["additional"] = append(crdMap["additional"], crd)
+		}
+	}
+
+	return crdMap
+}
```
Comment (brudnak) [inline tests/v2/validation/crdgeneration/crd_generation.go:52]: Just an idea for this section for less `else if` statements. Also, if a group ever needed to be added or removed it could just be added to or removed from the slice. Just an optional idea

```go
func mapCRD(crdList []string) map[string][]string {
	groups := []string{groupManagement, groupCatalog, groupUI, groupCluster, groupProject, groupRKE, groupProvisioning, groupFleet}
	crdMap := make(map[string][]string)

	for _, group := range groups {
		crdMap[group] = []string{}
	}

	crdMap["additional"] = []string{}

	for _, crd := range crdList {
		found := false
		for _, group := range groups {
			if strings.Contains(crd, group) {
				crdMap[group] = append(crdMap[group], crd)
				found = true
				break
			}
		}
		if !found {
			crdMap["additional"] = append(crdMap["additional"], crd)
		}
	}

	return crdMap
}

```

**Review by brudnak — CHANGES_REQUESTED** [review]
Just 1 comment / optional suggestion. Otherwise, lgtm

File: tests/v2/validation/crdgeneration/crd_generation.go (hunk 1 above)
Comment (anupama2501) [reply tests/v2/validation/crdgeneration/crd_generation.go]: yeah I think both has its pros and cons. using multiple if else might give us better readability but as you rightly pointed it adds complexity if we ever need to add/remove a group and more code duplication. Updated the PR to include the suggestion.

File: tests/framework/extensions/crds/listcrd.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package crds
+
+import (
+	v1 "github.com/rancher/rancher/tests/framework/clients/rancher/v1"
+)
+
+func GetCRDs(steveClient *v1.Client, clusterID string) (*v1.SteveCollection, error) {
```
Comment (igomez06) [inline tests/framework/extensions/crds/listcrd.go:7]: Question did the existing one not work for you? https://github.com/rancher/rancher/blob/release/v2.7/tests/framework/extensions/kubeapi/customresourcedefinitions/list.go
Or you needed it to be a v1 Steve call?

Comment (igomez06) [reply tests/framework/extensions/crds/listcrd.go:7]: Based on the answer the previous question add a comment if we're keeping this.

File: tests/framework/extensions/kubectl/apply.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	if err != nil {
+		return "", err
+	}
+
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "rancher-installer",
+		},
+	}
+	_, err = downClient.Resource(corev1.SchemeGroupVersion.WithResource("serviceaccounts")).Namespace(Namespace).Create(context.TODO(), ext_unstructured.MustToUnstructured(sa), metav1.CreateOptions{})
```
Comment (igomez06) [inline tests/framework/extensions/kubectl/apply.go:61]: Maybe for these kubectl calls, we can consolidated the resource creation into a function, and the apply and explain calls just receive the kubectl call. Also do we need to crate the service accounts and cluster role bindings.

File: tests/framework/extensions/kubectl/explain.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	if err != nil {
+		return "", err
+	}
+
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "rancher-installer2",
+		},
+	}
+	_, err = downClient.Resource(corev1.SchemeGroupVersion.WithResource("serviceaccounts")).Namespace(Namespace).Create(context.TODO(), ext_unstructured.MustToUnstructured(sa), metav1.CreateOptions{})
```
Comment (igomez06) [inline tests/framework/extensions/kubectl/explain.go:56]: See my comment for apply

File: tests/framework/extensions/kubectl/template.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+var (
+	importTimeout = int64(60 * 1)
+)
+
+// GetPodAndPodLogs gets the pod name and then capture the podLogs for that pod. 
+//It then returns the podlogs by using the helper GetPodLogs
+func GetPodAndPodLogs(client *rancher.Client, clusterID string) (string, error) {
+	steveClient := client.Steve
```
Comment (igomez06) [inline tests/framework/extensions/kubectl/template.go:46]: Don't we already have this func?

**Review by igomez06 — CHANGES_REQUESTED** [review]
looks great, please see my comments

File: tests/framework/extensions/crds/listcrd.go (hunk 2 above)
Comment (anupama2501) [reply tests/framework/extensions/crds/listcrd.go:7]: Wanted a steve call so added this extension. The other one is using dynamic client. So, added this extension.

File: tests/framework/extensions/kubectl/apply.go (hunk 3 above)
Comment (anupama2501) [reply tests/framework/extensions/kubectl/apply.go:61]: yes we need the sa but not cluster role bindings as we have to get the pod logs without the SA, we would get a forbidden error.

File: tests/framework/extensions/kubectl/template.go (hunk 5 above)
Comment (anupama2501) [reply tests/framework/extensions/kubectl/template.go:46]: We need the name of the pod for both kubectl apply and kubectl explain function. Instead of duplicating the code to get the pod name, added a function to get the pod name and then calling the existing GetPodLogs function.

**Review by caliskanugur — COMMENTED** [review]
PR looks really good to me; I've got some questions.

File: tests/framework/extensions/crds/listcrd.go (hunk 2 above)
Comment (anupama2501) [reply tests/framework/extensions/crds/listcrd.go:8]: @igomez06 checked with devs on this and its preferable to use kubernetes API as with the public API the preferable direction is to use kubernetes API and steve endpoints are used by UI mostly and for these extensions they might not update to v1 endpoints yet. Will update the PR to use Kubernetes API.

File: tests/framework/clients/rancher/config.go [hunk 6]
```diff
@@ -14,4 +14,5 @@ type Config struct {
 	CACerts       string `yaml:"caCerts" default:""`
 	ClusterName   string `yaml:"clusterName" default:""`
 	ShellImage    string `yaml:"shellImage" default:""`
+	ClusterAuth   bool  `yaml:"clusterAuth" default:"false"`
```
[the remark this answers is not in the document]
Comment (anupama2501) [reply tests/framework/clients/rancher/config.go:17]: ClusterAuth is enabled/disabled when ACE endpoints are enabled/disabled. This is only true when we are running the tests in a downstream cluster but not on a local cluster. When we have an RKE1 cluster created, ACE is enabled by default and the crds will be then generated in that downstream cluster but if we have an RKE2/K3s clusters, ACE is not enabled by default and depends on user configs. For this purpose, I added the ClusterAuth value here as it depends on user environments depending on the type of clusters they want to run the tests for.

File: tests/framework/extensions/crds/listcrd.go [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+	crds, err := steveClient.SteveType("apiextensions.k8s.io.customresourcedefinition").List(nil)
```
[the remark this answers is not in the document]
Comment (anupama2501) [reply tests/framework/extensions/crds/listcrd.go:10]: This won't be needed anymore as I will be removing this extension and making use of the listCRDs from the kybernetes API for which we already have an extension as thats the preferable path going forward.

File: tests/framework/extensions/kubectl/template.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+							Name: "config",
+							VolumeSource: corev1.VolumeSource{
+								EmptyDir: &corev1.EmptyDirVolumeSource{},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
```
[the remark this answers is not in the document]
Comment (anupama2501) [reply tests/framework/extensions/kubectl/template.go:60]: yeah definitely for any future tasks if we want to make use of the job template it would be easier. I can update the PR.

File: tests/framework/extensions/kubectl/template.go [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+	jobWatch, err := downClient.Resource(batchv1.SchemeGroupVersion.WithResource("jobs")).Namespace(Namespace).Watch(context.TODO(), metav1.ListOptions{
+		FieldSelector:  fields.OneTermEqualSelector("metadata.name", job.Name).String(),
+		TimeoutSeconds: &importTimeout,
+	})
+	if err != nil {
+		return err
+	}
+
+	_ = wait.WatchWait(jobWatch, func(event watch.Event) (bool, error) {
```
[the remark this answers is not in the document]
Comment (anupama2501) [reply tests/framework/extensions/kubectl/template.go:136]: No we don't want to. I forgot to update from my testing phase. I can change it back.

File: tests/v2/validation/crdgeneration/crd_generation.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
+package crdgeneration
```
[the remark this answers is not in the document]
Comment (anupama2501) [reply tests/v2/validation/crdgeneration/crd_generation.go:1]: yep agreed but the only thing I want to change here is instead of crd package we can change this to publicAPI instead and the generation_test.go which will be helpful for the publicAPI automation.

File: tests/framework/extensions/kubectl/template.go (hunk 8 above)
Comment (caliskanugur) [inline tests/framework/extensions/kubectl/template.go:60]: I don't think we need this; it's probably a missed thing. `NewJobTemplate` implemented.

File: tests/framework/extensions/workloads/template.go [hunk 11]
```diff
@@ -87,3 +91,33 @@ func NewDaemonSetTemplate(daemonsetName string, namespace string, template corev
...
+// NewJobTemplate is a constructor that creates a job template.
+func NewJobTemplate(jobName string, namespace string, template corev1.PodTemplateSpec) *batchv1.Job {
+	return &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: jobName,
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "",
```
Comment (caliskanugur) [inline tests/framework/extensions/workloads/template.go:104]: I've got a question for this empty string field; 

Could you verify on UI if it assigns this field an empty string while creating a workload/job?

I've got the same question for `ServiceAccountName` field.

File: tests/framework/extensions/workloads/template.go [hunk 12]
```diff
@@ -87,3 +91,33 @@ func NewDaemonSetTemplate(daemonsetName string, namespace string, template corev
...
+							Name: "config",
+							VolumeSource: corev1.VolumeSource{
+								EmptyDir: &corev1.EmptyDirVolumeSource{},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
```
Comment (caliskanugur) [inline tests/framework/extensions/workloads/template.go:121]: Also, as a follow-up question to the above; if it requires more things to add this function, we should/can be able to leave those parts something like this to use API's default:

```suggestion
	return &batchv1.Job{
		ObjectMeta: metav1.ObjectMeta{
			Name: jobName,
		},
		Spec: batchv1.JobSpec{
			Template: corev1.PodTemplateSpec{
				Spec: corev1.PodSpec{
					RestartPolicy:      "Never",
					Containers:         []corev1.Container{},
					Volumes: []corev1.Volume{
						{
							Name: "config",
							VolumeSource: corev1.VolumeSource{
								EmptyDir: &corev1.EmptyDirVolumeSource{},
							},
						},
					},
				},
			},
		},
	}
```

**Review by caliskanugur — COMMENTED** [review]
I've got some questions and comments. Looks good to me

File: tests/framework/extensions/kubectl/template.go (hunk 8 above)
Comment (anupama2501) [reply tests/framework/extensions/kubectl/template.go]: You're right. I forgot to update

File: tests/framework/extensions/workloads/template.go (hunk 11 above)
Comment (anupama2501) [reply tests/framework/extensions/workloads/template.go]: ServiceAccount is required as the user wont have permissions to run the commands otherwise.

File: tests/framework/extensions/workloads/template.go (hunk 12 above)
Comment (anupama2501) [reply tests/framework/extensions/workloads/template.go]: I have removed the Name field but left the ServiceAccountName as we would need the SA to run the kubectl command.

File: tests/framework/clients/rancher/config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (igomez06) [reply tests/framework/clients/rancher/config.go:17]: I think I understand, so this flag is if the cluster auth is enabled for the rancher instance?

File: tests/v2/validation/public_api/crdgeneration/crd_generation_test.go [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+			assert.Equal(crd.T(), crd.crds[groupProvisioning], value)
+			assert.Equal(crd.T(), len(crd.crds[groupProvisioning]), len(value))
+		case groupFleet:
+			assert.Equal(crd.T(), crd.crds[groupFleet], value)
+			assert.Equal(crd.T(), len(crd.crds[groupFleet]), len(value))
+		}
+	}
+}
+
+func (crd *CRDGenTestSuite) ValidateCRDDescription(crdList []string) {
```
Comment (igomez06) [inline tests/v2/validation/public_api/crdgeneration/crd_generation_test.go:99]: Can you update this so these functions are not exported.

**Review by igomez06 — COMMENTED** [review]
pretty solid, same feedback as how the rbac tests are structured.

File: tests/framework/clients/rancher/config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (anupama2501) [reply tests/framework/clients/rancher/config.go:17]: somewhat, but its not on rancher instance. Cluster auth crd is only applicable for RKE1 downstream clusters. The crd is not generated on the downstream If we have an RKE2/K3s clusters even when ACE is enabled.

File: tests/framework/extensions/crds/listcrd.go (hunk 2 above)
Comment (anupama2501) [reply tests/framework/extensions/crds/listcrd.go]: Spoke offline and it was agreed upon to remove clusterAuth altogether from rancher config and check if the cluster type if RKE1.

File: tests/framework/clients/rancher/config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (anupama2501) [reply tests/framework/clients/rancher/config.go]: Spoke offline and it was agreed upon to remove clusterAuth altogether from rancher config and check if the cluster type if RKE1.

File: tests/v2/validation/public_api/crdgeneration/crd_generation_test.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+	crd.clusterV1, err = kubeClient.Clusters(fleetNamespace).Get(context.TODO(), crd.clusterName, metav1.GetOptions{})
+	require.NoError(crd.T(), err)
+
+}
+
+func (crd *CRDGenTestSuite) sequentialTestCRD() {
+	crdsList, err := listCRDS(crd.client, crd.clusterID)
+	require.NoError(crd.T(), err)
+	crd.Run("Verify description fields of crds are non-empty", func() {
```
Comment (slickwarren) [inline tests/v2/validation/public_api/crdgeneration/crd_generation_test.go:77]: I think these run names are not representing the test that is running? looks like a copy/paste error

File: tests/v2/validation/public_api/crdgeneration/crd_generation_test.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+	readJson, err := os.ReadFile("../resources/crds.json")
+	require.NoError(crd.T(), err)
+	err = json.Unmarshal(readJson, &crd.crds)
+	require.NoError(crd.T(), err)
+
+	clusterType, err := clusters.NewClusterMeta(crd.client, crd.clusterName)
+	require.NoError(crd.T(), err)
+	//Change "local" to isLocal for Q4
+	if crd.clusterName != "local" && clusterType.Provider == clusters.KubernetesProviderRKE {
```
Comment (slickwarren) [inline tests/v2/validation/public_api/crdgeneration/crd_generation_test.go:59]: I think this would be easier to read as:
```
crd.clusterAuth = false
if crd.clusterName != "local" && clusterType.Provider == clusters.KubernetesProviderRKE && crd.cluster.LocalClusterAuthEndpoint != nil && crd.cluster.LocalClusterAuthEndpoint.Enabled:
  crd.clusterAuth = true
```

File: tests/v2/validation/public_api/crdgeneration/crd_generation_test.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+	crd.client = client
+	crd.clusterName = client.RancherConfig.ClusterName
+	require.NotEmptyf(crd.T(), crd.clusterName, "Cluster name to install should be set")
+	crd.clusterID, err = clusters.GetClusterIDByName(crd.client, crd.clusterName)
+	require.NoError(crd.T(), err, "Error getting cluster ID")
+	crd.cluster, err = crd.client.Management.Cluster.ByID(crd.clusterID)
+	require.NoError(crd.T(), err)
+
+	readJson, err := os.ReadFile("../resources/crds.json")
```
Comment (slickwarren) [inline tests/v2/validation/public_api/crdgeneration/crd_generation_test.go:51]: should this filepath be a const?

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+		desIndexEnd = len(logs) - 1
+	}
+
+	description := logs[desIndexStart+len("DESCRIPTION:") : desIndexEnd]
+	assert.NotContains(t, description, "<empty>")
+	log.Info(description)
+}
+
+func validateRoleCreation(t *testing.T, client *rancher.Client, clusterV1 *v1.Cluster, clusterID string) {
+	role, err := os.ReadFile("../resources/roleTemplate.json")
```
Comment (slickwarren) [inline tests/v2/validation/public_api/crdgeneration/crd_generation.go:125]: imo all these filepaths should be const, what do you all think?

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+func validateCRDDescription(t *testing.T, client *rancher.Client, clusterV1 *v1.Cluster, clusterID string) {
+
+	logs, err := kubectl.KubectlExplain(client, clusterV1, roleTemplate, clusterID)
+	require.NoError(t, err)
+	assert.NotEmpty(t, logs)
+	desIndexStart := strings.Index(logs, "DESCRIPTION:")
+	desIndexEnd := strings.Index(logs, "FIELDS")
+	if desIndexStart == -1 {
+		t.Fail()
+		t.Log("Pod logs do not contain description field")
```
Comment (slickwarren) [inline tests/v2/validation/public_api/crdgeneration/crd_generation.go:113]: won't the fail above this line make the log unable to run?

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+	for idx, crd := range crdCollection.Items {
+		crds[idx] = crd.GetName()
+	}
+	return crds, nil
+}
+
+func validateCRDList(t *testing.T, crdsList []string, crdMapPreUpgrade map[string][]string, clusterName string) {
+	crdMap := mapCRD(crdsList)
+
+	if clusterName == "local" {
```
Comment (slickwarren) [inline tests/v2/validation/public_api/crdgeneration/crd_generation.go:75]: should we add a 'warn' log or something if the cluster name is not local, so that we know no group assertions were checked?

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+
+	for _, group := range groups {
+		crdMap[group] = []string{}
+	}
+
+	crdMap["additional"] = []string{}
+
+	for _, crd := range crdList {
+		found := false
+		for _, group := range groups {
```
Comment (slickwarren) [inline tests/v2/validation/public_api/crdgeneration/crd_generation.go:44]: I think this could be more efficient, but if these lists are generally short then it should be fine as is.

File: tests/framework/extensions/workloads/template.go [hunk 21]
```diff
@@ -4,18 +4,22 @@ import (
...
 
 	appv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // NewContainer is a contructor that creates a container for a pod template i.e. corev1.PodTemplateSpec
-func NewContainer(containerName, image string, imagePullPolicy corev1.PullPolicy, volumeMounts []corev1.VolumeMount, envFrom []corev1.EnvFromSource) corev1.Container {
+func NewContainer(containerName, image string, imagePullPolicy corev1.PullPolicy, volumeMounts []corev1.VolumeMount, envFrom []corev1.EnvFromSource, command []string, securityContext *corev1.SecurityContext, args []string) corev1.Container {
```
Comment (slickwarren) [inline tests/framework/extensions/workloads/template.go:13]: this is more of a question for Ugur since he wrote this function; why can't we expect users to directly create their containers from corev1.Contianer? Otherwise I can see another 'createCluster' issue, with tons of parameters as we discover we need more for different test cases.

File: tests/framework/extensions/kubectl/apply.go [hunk 22]
```diff
@@ -0,0 +1,77 @@
+package kubectl
```
Comment (slickwarren) [inline tests/framework/extensions/kubectl/apply.go:1]: are these part of the CRD generation? if not, I think you could move these to a separate PR.

**Review by slickwarren — COMMENTED** [review]
some nits and questions, overall lgtm. Pending jenkins run/review. Also, is the kubectl stuff meant to be included in this PR?

**Review by igomez06 — CHANGES_REQUESTED** [review]
looks great, love the design. Some nitpicks similar to Caleb.

File: tests/v2/validation/public_api/crdgeneration/crd_generation_test.go (hunk 15 above)
Comment (anupama2501) [reply tests/v2/validation/public_api/crdgeneration/crd_generation_test.go:59]: I initially went with this style but later changed it to the current style for its brevity. IMO adding this in a single line would be difficult to read.  @igomez06 @caliskanugur any suggestions?

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go (hunk 19 above)
Comment (anupama2501) [reply tests/v2/validation/public_api/crdgeneration/crd_generation.go:75]: For the current release aka 2.7.7, we are only generating the crds for local cluster and these crds are only present in the local cluster. They not available in the downstream cluster. So, adding a warn log would not be appropriate here and also log message is not needed - updated rb.Run() logs the clusterName here: https://github.com/rancher/rancher/blob/8607ca927ecd4c55dc18ac36ec80d37ef8fd165e/tests/v2/validation/public_api/crdgeneration/crd_generation_test.go#L77

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go (hunk 20 above)
Comment (anupama2501) [reply tests/v2/validation/public_api/crdgeneration/crd_generation.go:44]: These are generally short so keeping them as is.

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go (hunk 18 above)
Comment (anupama2501) [reply tests/v2/validation/public_api/crdgeneration/crd_generation.go:113]: no the test will be failed but whatever is present after the t.Fail() will also be executed. So updated it to be t.Fatalf() which logs and fails the tests.

File: tests/framework/extensions/workloads/template.go (hunk 21 above)
Comment (anupama2501) [reply tests/framework/extensions/workloads/template.go:13]: do you have an example for this that you have in mind? IMO it will be easier to use a template instead of directly sending the data every time we make use of containers cc: @caliskanugur

File: tests/v2/validation/public_api/crdgeneration/crd_generation_test.go (hunk 15 above)
Comment (igomez06) [reply tests/v2/validation/public_api/crdgeneration/crd_generation_test.go:59]: I don't have a strong opinion, I would be ok with you implementation Anu

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+
+func validateCRDDescription(t *testing.T, client *rancher.Client, clusterV1 *v1.Cluster, clusterID string) {
+
+	logs, err := kubectl.KubectlExplain(client, clusterV1, roleTemplate, clusterID)
+	require.NoError(t, err)
+	assert.NotEmpty(t, logs)
+	desIndexStart := strings.Index(logs, "DESCRIPTION:")
+	desIndexEnd := strings.Index(logs, "FIELDS")
+	if desIndexStart == -1 {
+		t.Fatalf("Pod logs do not contain description field")
```
Comment (igomez06) [inline tests/v2/validation/public_api/crdgeneration/crd_generation.go:114]: would this be a drastic error? We should not use fatals, this would affect code coverage as well unless this is a breaking error. Maybe a forced require error so the test ends here.

Comment (anupama2501) [reply tests/v2/validation/public_api/crdgeneration/crd_generation.go:114]: I wouldn't say its a drastic error here. I can change it to require.Error(), but could you let me know why/how it would affect the code coverage ?

Comment (igomez06) [reply tests/v2/validation/public_api/crdgeneration/crd_generation.go:114]: Fatals are for extreme cases that will exit the program immediately, and not safety.  Doing so will affect our code coverage process, because we need a graceful exit to be able to write test reports. Especially in testing we should refrain from using fatals at all.

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+				assert.Equal(t, len(crdMapPreUpgrade[groupRKE]), len(value))
+			case groupProvisioning:
+				assert.Equal(t, crdMapPreUpgrade[groupProvisioning], value)
+				assert.Equal(t, len(crdMapPreUpgrade[groupProvisioning]), len(value))
+			case groupFleet:
+				assert.Equal(t, crdMapPreUpgrade[groupFleet], value)
+				assert.Equal(t, len(crdMapPreUpgrade[groupFleet]), len(value))
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (anupama2501) [reply tests/v2/validation/public_api/crdgeneration/crd_generation.go]: thanks for pointing that out. Updated.

File: tests/v2/validation/public_api/crdgeneration/crd_generation.go (hunk 18 above)
Comment (anupama2501) [reply tests/v2/validation/public_api/crdgeneration/crd_generation.go:97]: I tried a failed condition and noticed the cleanup does run and resources were removed.

**Review by slickwarren — COMMENTED** [review]
changes look good, Anu do we have a recent jenkins run?

**Review by igomez06 — APPROVED** [review]
LGTM

**Review by caliskanugur — APPROVED** [review]

**Review by brudnak — DISMISSED** [review]

**Comment by caliskanugur** [comment]
Could we get a rebase? Thank you!

**Review by slickwarren — DISMISSED** [review]
lgtm

**Comment by slickwarren** [comment]
actually, there is still a linter issue could you take a look @anupama2501

**Review by igomez06 — APPROVED** [review]
LGTM

**Review by caliskanugur — APPROVED** [review]
LGTM
